### DISCUSSION TITLE
CHAPTER20 RedisとJWTを利用する目的について

### DISCUSSION THREAD

**Discussion by cotta0322** [budougumi0617/go_todo_app#78] [Q&A]
まずは、執筆ありがとうございます。
とても参考にさせていただいてます！

早速質問です。
今回、CHAPTER20でRedisとJWTの２つを利用している目的が知りたく質問させていただきました。

JWTを利用した場合、Redisを使わなくても、本来はJWTの検証のみで認可されたリクエストか検証できると思います。
そして今回敢えてRedisでアクセストークンの管理を行っているのは有効期限に関わらず任意のタイミングでアクセストークンを無効にできるからだと思っています。

ただ、結局サーバー側（Redis）でJWTの管理を行うならば、そもそもJWTを用いるメリットがわかりません。
ただのランダム文字列で生成したAccessTokenでも構わない気がしています。

今回のようにサーバー側でTokenの管理を行うのに、それと合わせてJWTも利用するメリットがあれば伺いたいです。

**Comment by budougumi0617** [comment]
コメントありがとうございます。

Redisの利用目的はご記載のとおりです。JWTを使う目的としては例えば以下があると思います。

- 認証済みのユーザーに特定ロールを「認可」する
-（改ざん防止を施した）署名付きデータとして独自情報（Claim）を含める

これらをランダム文字列で実現するにはけっこう手の込んだ実装をしないといけないと思うのですがいかがでしょうか。

**Comment by cotta0322** [comment]
> 認証済みのユーザーに特定ロールを「認可」する
> -（改ざん防止を施した）署名付きデータとして独自情報（Claim）を含める

特定ロールや独自情報も、Token（ランダム文字列）と紐付けたプロパティとしてRedisで管理すると良いかなと思っています。
そうするとクライアント側とはToken（ランダム文字列）のやり取りしか発生しないため、そもそも改ざん等を考慮する必要もなくなります。
（これは従来からあるセッション管理に近いものかも）

ただ、本書のようにJWTを使ってもやれると思いますし、実現方法は複数あって、どれが良いかはプロダクトの状況や規模によりけりな気もしてきました....。
ここの議論は本書の趣旨ともずれてくると思いますので、もう少し自分でも考えてみようと思います。

ご丁寧に回答いただきありがとうございます。

**Reply by budougumi0617** [reply]
たしかに…！こちらこそご指摘ありがとうございます、私も勉強になります…!!
